File: frontend/src/utils/printPrice.ts
/**
 * What a finished print should sell for (fork).
 *
 * Upstream's finance module is a chargeback system — cost centres, budgets,
 * per-user wallets. It answers "who owes what for the machine time". A shop
 * asks the other question, and nothing upstream answers it, even though the
 * inputs are all on the archive row already: the filament `cost`, the
 * `energy_cost` the smart plug measured, and `print_time_seconds`.
 *
 * Kept as a pure function so the number can be tested without rendering, and
 * so it reacts to a settings change without a round trip.
 */

export interface PricingSettings {
  pricing_enabled: boolean;
  pricing_labour_per_hour: number;
  pricing_markup: number;
  pricing_floor: number;
}

/** Only the archive fields the price depends on. */
export interface PricedArchive {
  cost?: number | null;
  energy_cost?: number | null;
  print_time_seconds?: number | null;
}

export interface PrintPrice {
  /** Filament plus energy: what the archive actually measured. */
  materialCost: number;
  /** Labour rate times print hours. Zero when no rate is set. */
  labourCost: number;
  /** Everything the print cost to make, before markup. */
  unitCost: number;
  /** What to ask for it. */
  suggestedPrice: number;
  /**
   * Share of the selling price that is not cost, 0–1. Null when the price is
   * zero (nothing to take a share of), and negative when a markup below 1
   * prices the print under its own cost — which is worth showing, not hiding.
   */
  margin: number | null;
  /** True when the floor, not the markup, decided the price. */
  onFloor: boolean;
}

const SECONDS_PER_HOUR = 3600;

/**
 * Returns null when pricing is off, or when the print carries no cost
 * information at all — a suggestion built on nothing is worse than no
 * suggestion. A print with filament cost but no energy reading is priced with
 * energy treated as zero, because a plug without an energy meter is a normal
 * setup rather than missing data.
 */
export function priceForPrint(
  archive: PricedArchive,
  settings: Partial<PricingSettings> | null | undefined,
): PrintPrice | null {
  if (!settings?.pricing_enabled) return null;
  if (archive.cost == null && archive.energy_cost == null) return null;

  const materialCost = (archive.cost ?? 0) + (archive.energy_cost ?? 0);

  // Negative settings are treated as unset rather than refused: a stray minus
  // in a number field should not make the whole card show a nonsense price.
  const labourPerHour = Math.max(0, settings.pricing_labour_per_hour ?? 0);
  const hours = Math.max(0, archive.print_time_seconds ?? 0) / SECONDS_PER_HOUR;
  const labourCost = labourPerHour * hours;

  const unitCost = materialCost + labourCost;
  const markup = Math.max(0, settings.pricing_markup ?? 0);
  const floor = Math.max(0, settings.pricing_floor ?? 0);

  const marked = unitCost * markup;
  const suggestedPrice = Math.max(marked, floor);

  return {
    materialCost,
    labourCost,
    unitCost,
    suggestedPrice,
    margin: suggestedPrice > 0 ? (suggestedPrice - unitCost) / suggestedPrice : null,
    onFloor: suggestedPrice > marked,
  };
}

File: backend/app/services/print_price.py
"""Selling price for a finished print (fork).

Upstream's ``finance`` module is a chargeback system: cost centres, budgets,
per-user wallets, a print charged against a balance. It answers "who owes what
for the machine time", which is the question a makerspace or a school asks.

A shop asks the other question — what should this print sell for — and nothing
upstream answers it. The inputs are already there: every archive row carries the
filament ``cost`` and the ``energy_cost`` its smart plug measured, and
``print_time_seconds`` says how long the machine was busy.

The price itself is computed in the frontend (``utils/printPrice.ts``) so the
number reacts to a settings change without a round trip. What cannot be computed
there is the reference this module provides: the median unit cost across recent
prints, which is what makes a markup figure mean something. The list the browser
holds is paginated and filtered, so a median taken from it would be a median of
whatever page you were looking at.

Median, not mean: one resin-priced exotic filament or one nine-hour failure
would drag an average somewhere unhelpful, and the number exists to answer "what
does a typical print of mine cost".
"""

from datetime import datetime, timedelta, timezone

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.models.archive import PrintArchive


def median(values: list[float]) -> float | None:
    """Middle value, averaging the two middle ones for an even count.

    Returns None for an empty list rather than raising: "no prints yet" is a
    normal state for a fresh install, not an error.
    """
    if not values:
        return None
    ordered = sorted(values)
    mid = len(ordered) // 2
    if len(ordered) % 2 == 1:
        return ordered[mid]
    return (ordered[mid - 1] + ordered[mid]) / 2


def unit_cost(cost: float | None, energy_cost: float | None) -> float | None:
    """What one print cost to make, before any labour or markup.

    None only when neither figure is known — a print with filament cost but no
    energy reading is still worth counting, with energy treated as zero, because
    a plug without an energy meter is a common setup and dropping those prints
    would silently narrow the sample.
    """
    if cost is None and energy_cost is None:
        return None
    return (cost or 0.0) + (energy_cost or 0.0)


async def recent_unit_costs(db: AsyncSession, days: int = 90) -> list[float]:
    """Unit cost of every completed, non-deleted print in the window."""
    since = datetime.now(timezone.utc).replace(tzinfo=None) - timedelta(days=days)
    result = await db.execute(
        select(PrintArchive.cost, PrintArchive.energy_cost).where(
            PrintArchive.status == "completed",
            PrintArchive.deleted_at.is_(None),
            PrintArchive.created_at >= since,
        )
    )
    costs = []
    for cost, energy_cost in result.all():
        value = unit_cost(cost, energy_cost)
        # A zero-cost row means the filament was never priced, not that the
        # print was free; counting it would pull the median toward zero.
        if value is not None and value > 0:
            costs.append(value)
    return costs
